File: web/src/api/microFeed.ts
import { request } from './client';

export type MicroFeedStatus = 'draft' | 'published' | 'archived';
export type MicroFeedScript = 'cyrillic' | 'latin';
export type MicroFeedReaction = -1 | 0 | 1;

export interface DifficultWord {
  word: string;
  lemma: string;
  transcription: string;
  translationRu: string;
}

export interface MicroFeedItem {
  id: string;
  status: MicroFeedStatus;
  kind: 'news' | 'fact' | 'culture' | 'science' | 'fiction' | 'society' | 'book_excerpt';
  category:
    | 'history' | 'culture' | 'science' | 'fiction' | 'society' | 'news'
    | 'travel' | 'food' | 'sport' | 'music' | 'language';
  titleCyrillic: string;
  titleLatin: string;
  textCyrillic: string;
  textLatin: string;
  originalLanguage: string;
  originalScript: 'cyrillic' | 'latin' | 'translated';
  cefr: 'A1' | 'A2' | 'B1' | 'B2' | 'C1';
  tags: string[];
  difficultWords: DifficultWord[];
  imageUrl: string;
  audioUrl: string;
  sourceSlug: string;
  sourceImportId?: string;
  sourceTitle: string;
  sourceUrl: string;
  sourcePublishedAt: string | null;
  licenseCode: string;
  attributionText: string;
  bookId: string;
  chapterId: string;
  startPositionChar: number;
  bookTargetUrl: string;
  viewsCount: number;
  likesCount: number;
  dislikesCount: number;
  readMoreCount: number;
  commentsCount: number;
  reaction: MicroFeedReaction;
  hasEmbedding: boolean;
  publishedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface MicroFeedSource {
  slug: string;
  title: string;
  sourceKind: 'rss' | 'mediawiki' | 'manual';
  sourceUrl: string;
  language: string;
  rightsMode: 'reuse' | 'summary_only' | 'manual_review';
  licenseCode: string;
  attributionName: string;
  attributionUrl: string;
  enabled: boolean;
  lastSyncedAt: string | null;
}

export interface MicroFeedImport {
  id: string;
  sourceSlug: string;
  sourceTitle: string;
  externalId: string;
  title: string;
  sourceUrl: string;
  rawText: string;
  sourcePublishedAt: string | null;
  status: 'queued' | 'processed' | 'rejected';
  rejectionReason: string;
  createdAt: string;
}

export type MicroFeedItemDraft = Pick<MicroFeedItem,
  | 'kind' | 'category' | 'titleCyrillic' | 'titleLatin'
  | 'textCyrillic' | 'textLatin' | 'originalLanguage' | 'originalScript'
  | 'cefr' | 'tags' | 'difficultWords' | 'imageUrl' | 'audioUrl'
  | 'sourceSlug' | 'sourceTitle' | 'sourceUrl' | 'licenseCode'
  | 'attributionText' | 'bookId' | 'chapterId' | 'startPositionChar'
  | 'bookTargetUrl'
>;

/**
 * Идентификатор гостя выдаёт СЕРВЕР и подписывает своим ключом.
 *
 * Раньше его придумывал браузер: любой UUID принимался как новый читатель, и
 * лайк накручивался сменой строки в запросе. Здесь браузер только хранит то,
 * что ему выдали, вместе с лентой (`visitorToken` в ответе `/v1/micro-feed`).
 *
 * Прежний ключ localStorage намеренно другой: сохранённый самодельный UUID
 * сервер всё равно не примет, и подсовывать его в новый обмен незачем.
 */
const VISITOR_TOKEN_KEY = 'citavuk-micro-feed-visitor-token';

export function microFeedVisitorToken(): string {
  return localStorage.getItem(VISITOR_TOKEN_KEY) ?? '';
}

function rememberVisitorToken(token: string | undefined) {
  if (token) localStorage.setItem(VISITOR_TOKEN_KEY, token);
}

export async function getMicroFeed(exclude: string[], signal?: AbortSignal) {
  const query = new URLSearchParams({ limit: '8' });
  // Первый заход идёт без токена — сервер заведёт его и вернёт вместе с лентой.
  const token = microFeedVisitorToken();
  if (token) query.set('visitorToken', token);
  if (exclude.length > 0) query.set('exclude', exclude.slice(-80).join(','));

  const response = await request<{
    items: MicroFeedItem[];
    strategy: MicroFeedStrategy;
    preferences?: MicroFeedPreferences;
    visitorToken?: string;
  }>(`/v1/micro-feed?${query}`, { signal });
  rememberVisitorToken(response.visitorToken);
  return response;
}

export type MicroFeedStrategy = 'cold' | 'declared' | 'personalized';

/** Темы и уровень, названные читателем в анкете. */
export interface MicroFeedPreferences {
  categories: MicroFeedItem['category'][];
  cefr: MicroFeedItem['cefr'];
  /** Анкета пройдена. Пустой список тем — тоже ответ, а не «не спрашивали». */
  onboarded: boolean;
}

export const MICRO_FEED_LEVELS: MicroFeedItem['cefr'][] = ['A1', 'A2', 'B1', 'B2', 'C1'];

export function saveMicroFeedPreferences(
  categories: MicroFeedItem['category'][],
  cefr: MicroFeedItem['cefr'],
) {
  return request<MicroFeedPreferences>('/v1/micro-feed/preferences', {
    method: 'PUT',
    body: { visitorToken: microFeedVisitorToken(), categories, cefr },
  });
}

/** Карточки, отмеченные лайком: лайк работает ещё и закладкой. */
export async function getLikedMicroFeed(signal?: AbortSignal) {
  const query = new URLSearchParams({ visitorToken: microFeedVisitorToken() });
  const response = await request<{ items: MicroFeedItem[] }>(
    `/v1/micro-feed/liked?${query}`,
    { signal },
  );
  return response.items;
}

/**
 * Автоматический браузер: пререндер сборки, проверка ссылок, обход поисковика.
 *
 * Его показы и дочитывания в профиль идти не должны. Лента учится на поведении
 * читателя, а робот «дочитывает» каждую карточку за долю секунды и одинаково —
 * это не интерес, это шум, и в подборе он растворяет настоящие сигналы.
 * Пререндер сайта прогоняет ленту на каждой сборке, то есть шум был бы
 * регулярным.
 */
function automated(): boolean {
  return typeof navigator !== 'undefined' && navigator.webdriver === true;
}

export function recordMicroFeedInteraction(
  itemId: string,
  event: 'view' | 'like' | 'dislike' | 'reaction_cleared' | 'read_more_clicked' | 'quick_skip' | 'complete' | 'audio_play',
  dwellMs = 0,
) {
  if (automated()) return Promise.resolve();
  // Без токена действие учитывать не за кем. Он приходит вместе с лентой, то
  // есть к первому действию уже есть; отсутствие означает, что лента вообще не
  // загрузилась, и слать действие незачем.
  const token = microFeedVisitorToken();
  if (!token) return Promise.resolve();
  return request<void>(`/v1/micro-feed/${encodeURIComponent(itemId)}/interactions`, {
    method: 'POST',
    body: { visitorToken: token, event, dwellMs: Math.max(0, Math.round(dwellMs)) },
  });
}

export interface MicroFeedComment {
  id: string;
  itemId: string;
  userId: string;
  author: string;
  body: string;
  createdAt: string;
  /** Своя реплика: её можно удалить. Решает сервер, а не браузер. */
  mine: boolean;
}

/** Предел длины реплики. Тот же стоит на сервере и в базе. */
export const COMMENT_MAX = 600;

export const getMicroFeedComments = async (itemId: string, signal?: AbortSignal) =>
  (await request<{ items: MicroFeedComment[] }>(
    `/v1/micro-feed/${encodeURIComponent(itemId)}/comments`,
    { signal },
  )).items;

export const addMicroFeedComment = (itemId: string, body: string) =>
  request<MicroFeedComment>(`/v1/micro-feed/${encodeURIComponent(itemId)}/comments`, {
    method: 'POST', body: { body },
  });

export const deleteMicroFeedComment = (commentId: string) =>
  request<void>(`/v1/micro-feed/comments/${encodeURIComponent(commentId)}`, {
    method: 'DELETE',
  });

export const getAdminMicroFeedSources = () =>
  request<{ items: MicroFeedSource[]; generatorEnabled: boolean; embeddingsEnabled: boolean }>(
    '/v1/admin/micro-feed/sources',
  );

export const syncMicroFeedSource = (slug: string) =>
  request<{ found: number; saved: number }>(
    `/v1/admin/micro-feed/sources/${encodeURIComponent(slug)}/sync`,
    { method: 'POST', timeoutMs: 60_000 },
  );

export const getAdminMicroFeedImports = async (status = 'queued') =>
  (await request<{ items: MicroFeedImport[] }>(
    `/v1/admin/micro-feed/imports?status=${encodeURIComponent(status)}`,
  )).items;

export const generateMicroFeedItem = (id: string) =>
  request<MicroFeedItem>(`/v1/admin/micro-feed/imports/${encodeURIComponent(id)}/generate`, {
    method: 'POST', timeoutMs: 190_000,
  });

export const rejectMicroFeedImport = (id: string, reason: string) =>
  request<void>(`/v1/admin/micro-feed/imports/${encodeURIComponent(id)}/reject`, {
    method: 'POST', body: { reason },
  });

export const getAdminMicroFeedItems = async (status = '') =>
  (await request<{ items: MicroFeedItem[] }>(
    `/v1/admin/micro-feed/items?status=${encodeURIComponent(status)}`,
  )).items;

export const createMicroFeedItem = (body: MicroFeedItemDraft) =>
  request<MicroFeedItem>('/v1/admin/micro-feed/items', { method: 'POST', body });

export const updateMicroFeedItem = (id: string, body: MicroFeedItemDraft) =>
  request<MicroFeedItem>(`/v1/admin/micro-feed/items/${encodeURIComponent(id)}`, {
    method: 'PUT', body,
  });

export const publishMicroFeedItem = (id: string) =>
  request<MicroFeedItem>(`/v1/admin/micro-feed/items/${encodeURIComponent(id)}/publish`, {
    method: 'POST', timeoutMs: 60_000,
  });

export const archiveMicroFeedItem = (id: string) =>
  request<void>(`/v1/admin/micro-feed/items/${encodeURIComponent(id)}/archive`, { method: 'POST' });

export const deleteMicroFeedItem = (id: string) =>
  request<void>(`/v1/admin/micro-feed/items/${encodeURIComponent(id)}`, { method: 'DELETE' });

File: frontend/lib/models/micro_feed.dart
/// Карточка Вукотока — короткий сербский текст на один экран.
class MicroFeedItem {
  final String id;
  final String category;
  final String titleCyrillic;
  final String titleLatin;
  final String textCyrillic;
  final String textLatin;
  final String cefr;
  final List<String> tags;
  final List<DifficultWord> difficultWords;
  final String imageUrl;
  final String sourceTitle;
  final String sourceUrl;
  final String attributionText;
  final int likesCount;
  final int dislikesCount;
  final int commentsCount;

  /// −1, 0 или 1. Приходит от сервера для этого читателя.
  final int reaction;

  const MicroFeedItem({
    required this.id,
    required this.category,
    required this.titleCyrillic,
    required this.titleLatin,
    required this.textCyrillic,
    required this.textLatin,
    required this.cefr,
    required this.tags,
    required this.difficultWords,
    required this.imageUrl,
    required this.sourceTitle,
    required this.sourceUrl,
    required this.attributionText,
    required this.likesCount,
    required this.dislikesCount,
    required this.commentsCount,
    required this.reaction,
  });

  String title(bool cyrillic) => cyrillic ? titleCyrillic : titleLatin;
  String text(bool cyrillic) => cyrillic ? textCyrillic : textLatin;

  factory MicroFeedItem.fromJson(Map<String, dynamic> j,
      {required String imageBase}) {
    String s(String key) => (j[key] ?? '').toString();
    // Адрес картинки сервер отдаёт своей ручкой («/v1/micro-feed/…/image»):
    // прямая ссылка на чужой сайт рассказала бы ему, кто открыл карточку.
    final image = s('imageUrl');
    return MicroFeedItem(
      id: s('id'),
      category: s('category'),
      titleCyrillic: s('titleCyrillic'),
      titleLatin: s('titleLatin'),
      textCyrillic: s('textCyrillic'),
      textLatin: s('textLatin'),
      cefr: s('cefr'),
      tags: [for (final t in (j['tags'] as List?) ?? const []) t.toString()],
      difficultWords: [
        for (final w in (j['difficultWords'] as List?) ?? const [])
          DifficultWord.fromJson(w as Map<String, dynamic>),
      ],
      imageUrl: image.isEmpty
          ? ''
          : (image.startsWith('http') ? image : '$imageBase$image'),
      sourceTitle: s('sourceTitle'),
      sourceUrl: s('sourceUrl'),
      attributionText: s('attributionText'),
      likesCount: (j['likesCount'] as num?)?.toInt() ?? 0,
      dislikesCount: (j['dislikesCount'] as num?)?.toInt() ?? 0,
      commentsCount: (j['commentsCount'] as num?)?.toInt() ?? 0,
      reaction: (j['reaction'] as num?)?.toInt() ?? 0,
    );
  }
}

class DifficultWord {
  final String word;
  final String lemma;
  final String transcription;
  final String translationRu;

  const DifficultWord({
    required this.word,
    required this.lemma,
    required this.transcription,
    required this.translationRu,
  });

  factory DifficultWord.fromJson(Map<String, dynamic> j) => DifficultWord(
        word: (j['word'] ?? '').toString(),
        lemma: (j['lemma'] ?? '').toString(),
        transcription: (j['transcription'] ?? '').toString(),
        translationRu: (j['translationRu'] ?? '').toString(),
      );
}

/// Темы и уровень, названные читателем в анкете.
class MicroFeedPreferences {
  final List<String> categories;
  final String cefr;

  /// Анкета пройдена. Пустой список тем — тоже ответ, а не «не спрашивали».
  final bool onboarded;

  /// Уровень взят с аккаунта, где он задан один раз для всего приложения.
  /// Тогда анкета про уровень не спрашивает: человек уже ответил, и повторный
  /// вопрос выглядит так, будто его не услышали.
  final bool levelFromAccount;

  const MicroFeedPreferences({
    required this.categories,
    required this.cefr,
    required this.onboarded,
    this.levelFromAccount = false,
  });

  factory MicroFeedPreferences.fromJson(Map<String, dynamic> j) =>
      MicroFeedPreferences(
        categories: [
          for (final c in (j['categories'] as List?) ?? const []) c.toString(),
        ],
        cefr: (j['cefr'] ?? 'B1').toString(),
        onboarded: j['onboarded'] == true,
        levelFromAccount: j['levelFromAccount'] == true,
      );
}

/// Реплика в обсуждении карточки.
class MicroFeedComment {
  final String id;
  final String author;
  final String body;
  final DateTime? createdAt;

  /// Своя реплика — её можно удалить.
  final bool mine;

  const MicroFeedComment({
    required this.id,
    required this.author,
    required this.body,
    required this.createdAt,
    required this.mine,
  });

  factory MicroFeedComment.fromJson(Map<String, dynamic> j) {
    String s(String key) => (j[key] ?? '').toString();
    return MicroFeedComment(
      id: s('id'),
      author: s('author').isEmpty ? 'Читатель' : s('author'),
      body: s('body'),
      createdAt: DateTime.tryParse(s('createdAt'))?.toLocal(),
      mine: j['mine'] == true,
    );
  }
}

class MicroFeedPage {
  final List<MicroFeedItem> items;
  final String strategy;
  final MicroFeedPreferences? preferences;

  const MicroFeedPage({
    required this.items,
    required this.strategy,
    required this.preferences,
  });

  factory MicroFeedPage.fromJson(Map<String, dynamic> j,
          {required String imageBase}) =>
      MicroFeedPage(
        items: [
          for (final raw in (j['items'] as List?) ?? const [])
            MicroFeedItem.fromJson(raw as Map<String, dynamic>,
                imageBase: imageBase),
        ],
        strategy: (j['strategy'] ?? 'cold').toString(),
        preferences: j['preferences'] is Map<String, dynamic>
            ? MicroFeedPreferences.fromJson(
                j['preferences'] as Map<String, dynamic>)
            : null,
      );
}

/// Темы ленты. Порядок и подписи те же, что на сайте.
const microFeedCategories = <String, String>{
  'history': 'История',
  'culture': 'Культура',
  'science': 'Наука',
  'fiction': 'Литература',
  'society': 'Общество',
  'news': 'Новости',
  'travel': 'Путешествия',
  'food': 'Еда',
  'sport': 'Спорт',
  'music': 'Музыка',
  'language': 'Про язык',
};

const microFeedCategoryHints = <String, String>{
  'history': 'Балканы, короли, войны',
  'culture': 'Кино, обычаи, искусство',
  'science': 'Открытия и объяснения',
  'fiction': 'Отрывки из книг',
  'society': 'Как живут люди',
  'news': 'Что происходит сейчас',
  'travel': 'Куда съездить на Балканах',
  'food': 'Кухня, застолье, рецепты',
  'sport': 'Баскетбол, теннис, футбол',
  'music': 'Песни, группы, труба',
  'language': 'Слова, выражения, тонкости',
};

const microFeedLevels = <String, String>{
  'A1': 'Первые слова',
  'A2': 'Простые фразы',
  'B1': 'Читаю с переводчиком',
  'B2': 'Читаю почти свободно',
  'C1': 'Свободно',
};
